### PR TITLE
refactor(blooms): Add RPC service for bloom-planner

### PR DESCRIPTION
**Pull request by salvacorts**
**What this PR does / why we need it**:
This PR adds a new RPC service implemented by the planner to communicate the planner and the builder. The implementation is pretty similar top the [query scheduler][1].

**Special notes for your reviewer**:
- Since we haven't built yet the builder, I implemented a with a builder mock.
- I'm passing TSDB indentifiers and block refs as strings. We can reconsider defining protos for them in the future.

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

[1]: https://github.com/grafana/loki/blob/4c88be0ef2310406206c02d7747bf73b51c95bda/pkg/scheduler/scheduler.go#L420

### CODE REVIEW

File: pkg/bloombuild/planner/planner_test.go [hunk 1]
```diff
@@ -319,3 +332,163 @@ func Test_blockPlansForGaps(t *testing.T) {
...
+	require.NoError(t, err)
+	require.Eventually(t, func() bool {
+		return planner.State() == services.Running
+	}, 15*time.Second, 10*time.Millisecond)
+	defer func() {
+		planner.StopAsync()
+		require.Eventually(t, func() bool {
+			return planner.State() != services.Running
+		}, 1*time.Minute, 100*time.Millisecond)
+	}()
```
Comment (chaudum) [inline pkg/bloombuild/planner/planner_test.go:395]: You can use functions from the `services` package for starting and stopping:


```go
err := services.StartAndAwaitRunning(context.Background(), planner)
require.NoError(t, err)
t.Cleanup(func() {
    err := services.StopAndAwaitTerminated(context.Background(), planner)
    require.NoError(t, err)
})
```

**Review by chaudum — APPROVED** [review]

File: pkg/bloombuild/planner/planner_test.go (hunk 1 above)
Comment (salvacorts) [reply pkg/bloombuild/planner/planner_test.go:395]: Oh nice, TIL! Thank you 👍

File: pkg/bloombuild/planner/planner.go [hunk 2]
```diff
@@ -92,13 +101,23 @@ func New(
 	return p, nil
 }
 
-func (p *Planner) starting(_ context.Context) (err error) {
+func (p *Planner) starting(ctx context.Context) (err error) {
+	if err := services.StartManagerAndAwaitHealthy(ctx, p.subservices); err != nil {
+		return fmt.Errorf("error starting planner subservices: %w", err)
+	}
+
 	p.metrics.running.Set(1)
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloombuild/planner/planner.go:109]: We use it for other components and in the original bloom compactor.
Even though you may use a k82 metric to check if the pod is running:
- That won't work in non- k82 envs
- This is set to 1 only when the component is actually ready to start working

So I think that's the usefulness of this metric.

File: pkg/bloombuild/planner/planner.go [hunk 3]
```diff
@@ -108,20 +127,31 @@ func (p *Planner) running(ctx context.Context) error {
...
+		case <-planningTicker.C:
 			level.Info(p.logger).Log("msg", "starting bloom build iteration")
 			if err := p.runOne(ctx); err != nil {
 				level.Error(p.logger).Log("msg", "bloom build iteration failed", "err", err)
 			}
+
+		case <-inflightTasksTicker.C:
+			p.pendingTasksMu.Lock()
+			inflight := len(p.pendingTasks)
+			p.pendingTasksMu.Unlock()
```
[the remark this answers is not in the document]
Comment (salvacorts) [reply pkg/bloombuild/planner/planner.go]: Good idea. Done

**Pull request merged**